src/iter/step_by.rs

```diff
@@ -0,0 +1,131 @@
+use super::plumbing::*;
+use super::*;
+use crate::math::div_round_up;
+use std::iter;
+use std::usize;
+
+/// `StepBy` is an iterator that skip `n` element between each yield. where `n` is the given step
+/// This struct is created by the [`step_by()`] method on [`IndexedParallelIterator`]
+///
+/// [`step_by()`]: trait.IndexedParallelIterator.html#method.step_by
+/// [`IndexedParallelIterator`]: trait.IndexedParallelIterator.html
+#[cfg(step_by)]
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+#[derive(Debug, Clone)]
+pub struct StepBy<I: IndexedParallelIterator> {
+    base: I,
+    step: usize,
+}
+
+#[cfg(step_by)]
+impl<I> StepBy<I>
+    where
+        I: IndexedParallelIterator,
+{
+    /// Create a new `StepBy` iterator.
+    pub(super) fn new(base: I, step: usize) -> Self {
+        StepBy { base, step }
+    }
+}
+
+#[cfg(step_by)]
+impl<I> ParallelIterator for StepBy<I>
+    where
+        I: IndexedParallelIterator,
+{
+    type Item = I::Item;
+
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result
+        where
+            C: UnindexedConsumer<Self::Item>, {
+        bridge(self, consumer)
+    }
+
+    fn opt_len(&self) -> Option<usize> {
+        Some(self.len())
+    }
+}
+
+#[cfg(step_by)]
+impl<I> IndexedParallelIterator for StepBy<I>
+    where
+        I: IndexedParallelIterator,
+{
+    fn drive<C: Consumer<Self::Item>>(self, consumer: C) -> C::Result {
+        bridge(self, consumer)
+    }
+
+    fn len(&self) -> usize {
+        div_round_up(self.base.len(), self.step)
+    }
+
+    fn with_producer<CB>(self, callback: CB) -> CB::Output
+        where
+            CB: ProducerCallback<Self::Item>,
+    {
+        return self.base.with_producer(Callback {callback, step: self.step});
+
+        struct Callback<CB> {
+            callback: CB,
+            step: usize,
+        }
+
+        impl<T, CB> ProducerCallback<T> for Callback<CB>
+            where
+                CB: ProducerCallback<T>,
+        {
+            type Output = CB::Output;
+            fn callback<P>(self, base: P) -> CB::Output
+                where
+                    P: Producer<Item = T>,
+            {
+                let producer = StepByProducer { base, step: self.step };
+                self.callback.callback(producer)
+            }
+        }
+    }
+}
+
+/// ////////////////////////////////////////////////////////////////////////
+/// Producer implementation
+
+#[cfg(step_by)]
+struct StepByProducer<P> {
+    base: P,
+    step: usize,
+}
+
+#[cfg(step_by)]
+impl<P> Producer for StepByProducer<P>
+    where
+        P: Producer,
+{
+    type Item = P::Item;
+    type IntoIter = iter::StepBy<P::IntoIter>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.base.into_iter().step_by(self.step)
+    }
+
+    fn split_at(self, index: usize) -> (Self, Self) {
+        let (left, right) = self.base.split_at(index * self.step);
```

Comment:
I think this will fail if the step does not evenly divide the length, because `index * self.step` will be greater than the total length. You can reference `ChunksProducer::split_at` to see how to limit this.